composer/callbacks/checkpoint_saver.py

```diff
@@ -411,14 +415,28 @@ def _save_checkpoint(self, state: State, logger: Logger):
                 remote_prefix = state.sharded_ckpt_prefix_dir
                 remote_file_name = os.path.join(
                     pathlib.Path(remote_file_name).parent, remote_prefix,
-                    pathlib.Path(remote_file_name).name)
+                    _TORCH_DISTRIBUTED_CHECKPOINTS_FILENAME)
                 remote_file_name = format_name_with_dist_and_time(remote_file_name, state.run_name, state.timestamp)
+
+                # Upload metadata file.
+                if dist.get_global_rank() == 0:
+                    metadata_local_file_name = os.path.join(
+                        Path(saved_path).parent,
+                        _TORCH_DISTRIBUTED_CHECKPOINTS_METADATA_FILENAME
+                        )
+                    metadata_local_file_name = format_name_with_dist_and_time(metadata_local_file_name, state.run_name, state.timestamp)
+                    metadata_remote_file_name = os.path.join(
+                        Path(remote_file_name).parent,
+                        _TORCH_DISTRIBUTED_CHECKPOINTS_METADATA_FILENAME)
+                    metadata_remote_file_name = format_name_with_dist_and_time(metadata_remote_file_name, state.run_name, state.timestamp)
+                    logger.upload_file(remote_file_name=metadata_remote_file_name, 	file_path=metadata_local_file_name, overwrite=self.overwrite)
```

Comment:
just checking my understanding. only rank 0 needs to upload the metadata file because it is putting into a remote folder which will ultimately contain one file for each ranks checkpoint + one metadata file, whereas locally every rank needs to save the metadata file because they all have different folders. is that correct?

Reply:
so all ranks save to the same folder, so there is only one metadata file, but all ranks participate in writing it's content. The FileSystemManager does a big gather/reduce on the metadata from the other ranks and then dumps the result ot a file. But yeah we just need one of the ranks to upload the file to object store